arangod/IResearch/IResearchAqlAnalyzer.cpp

```diff
@@ -0,0 +1,516 @@
+    }
+  }
+  return nullptr;
+}
+}  // namespace
+
+namespace arangodb {
+namespace iresearch {
+
+/*static*/ bool AqlAnalyzer::normalize_vpack(const irs::string_ref& args, std::string& out) {
+  auto const slice = arangodb::iresearch::slice(args);
+  VPackBuilder builder;
+  if (normalize_slice(slice, builder)) {
+    out.resize(builder.slice().byteSize());
+    std::memcpy(&out[0], builder.slice().begin(), out.size());
+    return true;
+  }
+  return false;
+}
+
+/*static*/ bool AqlAnalyzer::normalize_json(const irs::string_ref& args,
+                                                    std::string& out) {
+  auto src = VPackParser::fromJson(args);
+  VPackBuilder builder;
+  if (normalize_slice(src->slice(), builder)) {
+    out = builder.toString();
+    return true;
+  }
+  return false;
+}
+
+/*static*/ irs::analysis::analyzer::ptr AqlAnalyzer::make_vpack(irs::string_ref const& args) {
+  Options options;
+  auto const slice = arangodb::iresearch::slice(args);
+  return make_slice(slice);
+}
+
+
+/*static*/ irs::analysis::analyzer::ptr AqlAnalyzer::make_json(irs::string_ref const& args) {
+  Options options;
+  auto builder = VPackParser::fromJson(args);
+  return make_slice(builder->slice());
+}
+
+AqlAnalyzer::AqlAnalyzer(Options const& options)
+  : irs::analysis::analyzer(irs::type<AqlAnalyzer>::get()),
+    _options(options),
+    _query(new CalculationQueryContext(arangodb::DatabaseFeature::getCalculationVocbase())),
+    _itemBlockManager(&_resourceMonitor, SerializationFormat::SHADOWROWS), 
+    _engine(0, *_query, _itemBlockManager,
+            SerializationFormat::SHADOWROWS, nullptr) {
+  TRI_ASSERT(validateQuery(_options.queryString,
+                           arangodb::DatabaseFeature::getCalculationVocbase())
+                 .ok());
+}
+
+bool AqlAnalyzer::next() {
+  do {
+    if (_queryResults != nullptr) {
+      while (_queryResults->numRows() > _resultRowIdx) {
+        AqlValue const& value =
+            _queryResults->getValueReference(_resultRowIdx++, _engine.resultRegister());
+        if (value.isString() || (value.isNull(true) && _options.keepNull)) {
+          if (value.isString()) {
+            _term.value = arangodb::iresearch::getBytesRef(value.slice());
+          } else {
+            _term.value = irs::bytes_ref::EMPTY;
+          }
+          _inc.value = _nextIncVal;
+          _nextIncVal = !_options.collapsePositions;
+          return true;
+        }
+      }
+    }
+    if (_executionState == ExecutionState::HASMORE) {
+      _executionState = ExecutionState::DONE;  // set to done to terminate in case of exception
+      _resultRowIdx = 0;
+      _queryResults = nullptr;
+      try {
+        std::tie(_executionState, _queryResults) = _engine.getSome(_options.batchSize);
+        TRI_ASSERT(_executionState != ExecutionState::WAITING);
+      } catch (basics::Exception const& e) {
+        LOG_TOPIC("b0026", ERR, iresearch::TOPIC)
```

Comment:
IMHO in user-defined queries we should not log errors to the logfile, but return the error to the caller, so the caller can handle it.
Otherwise the DBAs may see logfile errors but will not have any context about which query triggered it.

Reply:
Meaningful error will be generated during query validation phase. But if query passed validation and later this broke - we for sure should notify DBAs. And the context is here: "error executing calculation query" :) So log reader will know for sure - error is related to analyzers. I`ll add query string to log to make message more specific.